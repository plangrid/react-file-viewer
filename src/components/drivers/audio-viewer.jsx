// Copyright (c) 2017 PlanGrid, Inc.

import React, { Component } from 'react';

import 'styles/video.scss';
import Loading from '../loading';

class AudioViewer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
    };
  }

  onCanPlay() {
    this.setState({ loading: false });
  }

  renderLoading() {
    if (this.state.loading) {
      return <Loading />;
    }
    return null;
  }

  render() {
    const visibility = this.state.loading ? 'hidden' : 'visible';
    return (
      <div className="pg-driver-view">
        <div className="video-container">
          {this.renderLoading()}
          <audio
            style={{ visibility }}
            controls
            onCanPlay={e => this.onCanPlay(e)}
            src={this.props.filePath}
          >
            Video playback is not supported by your browser.
          </audio>
        </div>
      </div>
    );
  }
}

export default AudioViewer;
