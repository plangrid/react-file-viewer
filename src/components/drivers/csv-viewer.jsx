// Copyright (c) 2017 PlanGrid, Inc.

import React, { Component } from 'react';

import ReactDataGrid from 'react-data-grid';
import CSV from 'comma-separated-values';

class CsvViewer extends Component {

  static parse(data) {
    const rows = [];
    const columns = [];

    new CSV(data).forEach((array) => {
      if (columns.length < 1) {
        array.forEach((cell, idx) => {
          columns.push({
            key: `key-${idx}`,
            name: cell,
            resizable: true,
            sortable: true,
            filterable: true,
          });
        });
      } else {
        const row = {};
        array.forEach((cell, idx) => {
          row[`key-${idx}`] = cell;
        });
        rows.push(row);
      }
    });

    return { rows, columns };
  }

  constructor(props) {
    super(props);
    this.state = CsvViewer.parse(props.data);
  }

  componentWillReceiveProps(nextProps) {
    this.setState(CsvViewer.parse(nextProps.data));
  }

  render() {
    const { rows, columns } = this.state;
    return (
      <ReactDataGrid
        columns={columns}
        rowsCount={rows.length}
        rowGetter={i => rows[i]}
        minHeight={this.props.height || 650}
      />
    );
  }
}

export default CsvViewer;
