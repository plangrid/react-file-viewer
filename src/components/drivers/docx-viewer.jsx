// Copyright (c) 2017 PlanGrid, Inc.

import React, { Component } from 'react';
import mammoth from 'mammoth';

import 'styles/docx.scss';
import Loading from '../loading';

export default class extends Component {
  componentDidMount() {
    const jsonFile = new XMLHttpRequest();
    jsonFile.open('GET', this.props.filePath, true);
    jsonFile.send();
    jsonFile.responseType = 'arraybuffer';
    jsonFile.onreadystatechange = () => {
      if (jsonFile.readyState === 4 && jsonFile.status === 200) {
        mammoth.convertToHtml(
          { arrayBuffer: jsonFile.response },
          { includeDefaultStyleMap: true },
        )
        .then((result) => {
          const docEl = document.createElement('div');
          docEl.className = 'document-container';
          docEl.innerHTML = result.value;
          document.getElementById('docx').innerHTML = docEl.outerHTML;
        })
        .catch((a) => {
          console.log('alexei: something went wrong', a);
        })
        .done();
      }
    };
  }

  render() {
    return (
      <div id="docx">
        <Loading />
      </div>);
  }
}
