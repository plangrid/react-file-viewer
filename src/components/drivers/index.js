// Copyright (c) 2017 PlanGrid, Inc.

export { default as CsvViewer } from './csv-viewer';
export { default as Photo360Viewer } from './photo360-viewer';
export { default as PDFViewer } from './pdf-viewer';
export { default as DocxViewer } from './docx-viewer';
export { default as VideoViewer } from './video-viewer';
export { default as XlsxViewer } from './xlsx-viewer';
export { default as XBimViewer } from './xbim-viewer';
export { default as UnsupportedViewer } from './unsupported-viewer';
export { default as PhotoViewer } from './photo-viewer';
export { default as PhotoViewerWrapper } from './photo-viewer-wrapper';
export { default as AudioViewer } from './audio-viewer';
