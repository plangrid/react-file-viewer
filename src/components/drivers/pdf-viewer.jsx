// Copyright (c) 2017 PlanGrid, Inc.

import React from 'react';
import VisibilitySensor from 'react-visibility-sensor';
import { PDFJS } from 'pdfjs-dist/build/pdf.combined';
import 'pdfjs-dist/web/compatibility';

PDFJS.disableWorker = true;
const INCREASE_PERCENTAGE = 0.2;
const DEFAULT_SCALE = 1.1;

class PDFPage extends React.Component {

  constructor(props) {
    super(props);
    this.state = {};
    this.onChange = this.onChange.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    // we want to render/re-render in two scenarias
    // user scrolls to the pdf
    // user zooms in
    if (prevState.isVisible === this.state.isVisible && prevProps.zoom === this.props.zoom) return;
    if (this.state.isVisible) this.fetchAndRenderPage();
  }

  onChange(isVisible) {
    if (isVisible) this.setState({ isVisible });
  }

  fetchAndRenderPage() {
    const { pdf, index } = this.props;
    pdf.getPage(index).then(this.renderPage.bind(this));
  }

  renderPage(page) {
    const { containerWidth, zoom } = this.props;
    const calculatedScale = (containerWidth / page.getViewport(DEFAULT_SCALE).width);
    const scale = calculatedScale > DEFAULT_SCALE ? DEFAULT_SCALE : calculatedScale;
    const viewport = page.getViewport(scale + zoom);
    const { width, height } = viewport;

    const context = this.canvas.getContext('2d');
    this.canvas.width = width;
    this.canvas.height = height;

    page.render({
      canvasContext: context,
      viewport,
    });
  }

  render() {
    return (
      <div className="pdf-canvas">
        <VisibilitySensor onChange={this.onChange} partialVisibility >
          <canvas ref={node => this.canvas = node} width="670" height="870" />
        </VisibilitySensor>
      </div>
    );
  }
}

export default class PDFDriver extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      pdf: null,
      zoom: 0,
      percent: 0,
    };

    this.increaseZoom = this.increaseZoom.bind(this);
    this.reduceZoom = this.reduceZoom.bind(this);
    this.resetZoom = this.resetZoom.bind(this);
  }

  componentDidMount() {
    const { filePath } = this.props;
    const containerWidth = this.container.offsetWidth;
    PDFJS.getDocument(filePath, null, null, this.progressCallback.bind(this)).then((pdf) => {
      this.setState({ pdf, containerWidth });
    });
  }

  setZoom(zoom) {
    this.setState({
      zoom,
    });
  }

  progressCallback(progress) {
    const percent = ((progress.loaded / progress.total) * 100).toFixed();
    this.setState({ percent });
  }

  reduceZoom() {
    if (this.state.zoom === 0) return;
    this.setZoom(this.state.zoom - 1);
  }

  increaseZoom() {
    this.setZoom(this.state.zoom + 1);
  }

  resetZoom() {
    this.setZoom(0);
  }

  renderPages() {
    const { pdf, containerWidth, zoom } = this.state;
    if (!pdf) return null;
    const pages = Array.apply(null, { length: pdf.numPages });
    return pages.map((v, i) => (
      (<PDFPage
        index={i + 1}
        pdf={pdf}
        containerWidth={containerWidth}
        zoom={zoom * INCREASE_PERCENTAGE}
      />)
    ));
  }

  renderLoading() {
    if (this.state.pdf) return null;
    return (<div className="pdf-loading">LOADING ({this.state.percent}%)</div>);
  }

  render() {
    return (
      <div className="pdf-viewer-container">
        <div className="pdf-viewer" ref={node => this.container = node} >
          <div className="pdf-controlls-container">
            <div className="view-control" onClick={this.increaseZoom} >
              <i className="zoom-in" />
            </div>
            <div className="view-control" onClick={this.resetZoom}>
              <i className="zoom-reset" />
            </div>
            <div className="view-control" onClick={this.reduceZoom}>
              <i className="zoom-out" />
            </div>
          </div>
          {this.renderLoading()}
          {this.renderPages()}
        </div>
      </div>
    );
  }
}
