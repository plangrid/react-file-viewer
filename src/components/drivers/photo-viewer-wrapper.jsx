import React, { Component } from 'react';

import * as THREE from 'three';
import PhotoViewer from './photo-viewer';
import Photo360Viewer from './photo360-viewer';
import Loading from '../loading';

function getPhotoDriver(width, height, fileType) {
  if (fileType === 'jpg' && window.Math.abs((width / height) - 2) <= 0.01) {
    return Photo360Viewer;
  }
  return PhotoViewer;
}

export default class PhotoViewerWrapper extends Component {
  constructor(props) {
    super(props);

    this.state = {
      originalWidth: 0,
      originalHeight: 0,
      imageLoaded: false,
    };
  }

  componentDidMount() {
    this.loadFile();
  }

  componentDidUpdate() {
    this.loadFile();
  }

  loadFile() {
    // spike on using promises and a different loader or adding three js loading manager
    const loader = new THREE.TextureLoader();
    loader.crossOrigin = '';
    // load a resource
    loader.load(
      // resource URL
      this.props.filePath,
      // Function when resource is   loaded
      (texture) => {
        this.setState({
          originalWidth: texture.image.width,
          originalHeight: texture.image.height,
          imageLoaded: true,
          texture,
        });
      },
      (xhr) => {
        console.log(`${xhr.loaded / xhr.total * 100}% loaded`);
      },
      (xhr) => {
        console.log('An error happened', xhr);
      },
    );
  }

  render() {
    if (!this.state.imageLoaded) {
      return <Loading />;
    }
    const { originalWidth, originalHeight } = this.state;
    const PhotoDriver = getPhotoDriver(originalWidth, originalHeight, this.props.fileType);

    return (
      <PhotoDriver {...this.state} {...this.props} />
    );
  }
}
