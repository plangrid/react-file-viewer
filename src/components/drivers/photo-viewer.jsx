// Copyright (c) 2017 PlanGrid, Inc.

import React, { PureComponent } from 'react';

import 'styles/photo-viewer.scss';

export default class PhotoViewer extends PureComponent {


  componentDidMount() {
    this.insertImage();
    this.binded = this.onResize.bind(this)
    window.addEventListener("resize",this.binded)
  }

	componentWillUnmount() {
		window.removeEventListener("resize",this.binded)
  }

	onResize(){
    console.log("resize")
		const { originalWidth, originalHeight } = this.props;
		const imageDimensions = this.getImageDimensions.call(this, originalWidth, originalHeight);
		this.props.texture.image.style.width = `${imageDimensions.width}px`;
		this.props.texture.image.style.height = `${imageDimensions.height}px`;
		this.props.texture.image.setAttribute('class', 'photo');
  }

	componentWillReceiveProps(nextProps, nextContext) {
    const oldSrc = this.props.texture&&this.props.texture.image&&this.props.texture.image.src;
    const newSrc = nextProps.texture&&nextProps.texture.image&&nextProps.texture.image.src;
    if (oldSrc != newSrc){


    const removeCurrentImage = () => {
      const node = document.getElementById('pg-photo-container');
      while (node.hasChildNodes()) {
        node.removeChild(node.lastChild);
      }
    };

    removeCurrentImage();
    this.insertImage();
    } else{
      //no refresh
    }
	}


  getImageDimensions(originalWidth, originalHeight) {
    // Scale image to fit into viewer
    let imgHeight;
    let imgWidth;
    const { height: viewerHeight, width: viewerWidth } = this.props;

    if (originalHeight <= viewerHeight && originalWidth <= viewerWidth) {
      imgWidth = originalWidth;
      imgHeight = originalHeight;
    } else {
      const heightRatio = viewerHeight / originalHeight;
      const widthRatio = viewerWidth / originalWidth;
      if (heightRatio < widthRatio) {
        imgHeight = originalHeight * heightRatio;
        imgWidth = originalWidth * heightRatio;
      } else {
        imgHeight = originalHeight * widthRatio;
        imgWidth = originalWidth * widthRatio;
      }
    }

    return { height: imgHeight, width: imgWidth };
  }

  insertImage() {
    const { originalWidth, originalHeight } = this.props;
    const imageDimensions = this.getImageDimensions.call(this, originalWidth, originalHeight);

    this.props.texture.image.style.width = `${imageDimensions.width}px`;
    this.props.texture.image.style.height = `${imageDimensions.height}px`;
    this.props.texture.image.setAttribute('class', 'photo');
    document.getElementById('pg-photo-container').appendChild(this.props.texture.image);
  }

  render() {
    const containerStyles = {
      width: `${this.props.width}px`,
      height: `${this.props.height}px`,
    };

    return (
      <div style={containerStyles} className="photo-viewer-container" id="pg-photo-container" />
    );
  }
}
