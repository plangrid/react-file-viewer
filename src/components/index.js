// Copyright (c) 2017 PlanGrid, Inc.

module.exports = require('./file-viewer');
