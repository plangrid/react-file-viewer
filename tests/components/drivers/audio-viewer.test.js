// Copyright (c) 2017 PlanGrid, Inc.

import React from 'react';
import renderer from 'react-test-renderer';

import { AudioViewer } from 'components/drivers';

describe('AudioViewer', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(
      <AudioViewer filePath='fake/path' />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
