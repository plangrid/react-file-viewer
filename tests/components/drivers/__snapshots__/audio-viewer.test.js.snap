// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioViewer matches snapshot 1`] = `
<div
  className="pg-driver-view"
>
  <div
    className="video-container"
  >
    <div
      className="loading-container"
    >
      <span
        className="loading"
      />
    </div>
    <audio
      controls={true}
      onCanPlay={[Function]}
      src="fake/path"
      style={
        Object {
          "visibility": "hidden",
        }
      }
    >
      Video playback is not supported by your browser.
    </audio>
  </div>
</div>
`;
